runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingApplianceComputationConfigFetcher.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.beam.runners.dataflow.worker.windmill.WindmillServerStub;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.extensions.gcp.util.Transport;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.HashBasedTable;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Fetches computation config from Streaming Appliance. */
+@Internal
+@ThreadSafe
+public final class StreamingApplianceComputationConfigFetcher implements ComputationConfig.Fetcher {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(StreamingApplianceComputationConfigFetcher.class);
+
+  private final WindmillServerStub windmillServer;
+  private final ConcurrentHashMap<String, String> systemNameToComputationIdMap;
+
+  public StreamingApplianceComputationConfigFetcher(WindmillServerStub windmillServer) {
+    this.windmillServer = windmillServer;
+    this.systemNameToComputationIdMap = new ConcurrentHashMap<>();
+  }
+
+  private static Table<String, String, String> transformUserNameToStateFamilyByComputationId(
+      Windmill.GetConfigResponse response) {
+    // a row in the table is <ComputationId, TransformUserName, StateFamilyName>
+    Table<String, String, String> computationIdTransformUserNameStateFamilyNameTable =
+        HashBasedTable.create();
+    for (Windmill.GetConfigResponse.ComputationConfigMapEntry computationConfig :
+        response.getComputationConfigMapList()) {
+      for (Windmill.ComputationConfig.TransformUserNameToStateFamilyEntry entry :
+          computationConfig.getComputationConfig().getTransformUserNameToStateFamilyList()) {
+        computationIdTransformUserNameStateFamilyNameTable.put(
+            computationConfig.getComputationId(),
+            entry.getTransformUserName(),
+            entry.getStateFamily());
+      }
+    }
+
+    return computationIdTransformUserNameStateFamilyNameTable;
+  }
+
+  /** Deserialize {@link MapTask} and populate MultiOutputInfos in MapTask. */
+  private Optional<MapTask> deserializeAndFixMapTask(String serializedMapTask) {
+    try {
+      return Optional.of(Transport.getJsonFactory().fromString(serializedMapTask, MapTask.class));
+    } catch (IOException e) {
+      LOG.warn("Parsing MapTask failed: {}", serializedMapTask, e);
+    }
+    return Optional.empty();
+  }
+
+  @Override
+  public Optional<ComputationConfig> getConfig(String computationId) {
+    Preconditions.checkArgument(
+        !computationId.isEmpty(),
+        "computationId is empty. Cannot fetch computation config without a computationId.");
+
+    GetConfigResponse response =
+        windmillServer.getConfig(
```

Comment:
instead of taking in WindmillServer can you take in a functional interface matching this method?

That will help show that the full class isn't used and could simplify testing. It could help in the future break up windmillServer if we get rid of the jni class keeping it all tied together ATM.

Reply:
done